### PR TITLE
Feature/91-test-execution

### PR DESCRIPTION
**Pull request by AyTee001**
https://trello.com/c/ysVCg2tz/91-test-execution

### CODE REVIEW

File: backend/LeetWars.Builder/Services/XmlTestResultParserService.cs
Comment (pospyra) [inline backend/LeetWars.Builder/Services/XmlTestResultParserService.cs:1]: I think it would be better to split these methods into smaller parts (methods)

File: frontend/src/app/modules/online-editor/online-editor-page/online-editor-page.component.ts [hunk 1]
```diff
@@ -104,6 +132,50 @@ export class OnlineEditorPageComponent implements OnDestroy, OnInit {
...
+                        '{\r\n        throw new Exception("Exception!!!");\r\n    }\r\n}\r\n',
+            preloaded: null,
+            tests: 'using NUnit.Framework;\r\n\r\n[TestFixture]\r\npublic class Tests\r\n{\r\n    ' +
+                    'private Solution? _solutionClass;\r\n\r\n    [SetUp]\r\n    ' +
+                    'public void Setup()\r\n    {\r\n        _solutionClass = new Solution();\r\n    }\r\n\r\n    ' +
+                    '[Test]\r\n    public void IsPrime_InputIs1_ReturnFalse()\r\n    ' +
+                    '{\r\n        var result = _solutionClass.IsNumPrime(2);\r\n\r\n        ' +
+                    'Assert.IsFalse(result, "1 should not be prime");\r\n    }\r\n}',
+        };
+        // const codeRunRequest: CodeRunRequest = {
```
Comment (pospyra) [inline frontend/src/app/modules/online-editor/online-editor-page/online-editor-page.component.ts]: Please remove the commented code.

File: backend/LeetWars.Notifier/LeetWars.Notifier.WebAPI/Models/CodeRunResults.cs [hunk 2]
```diff
@@ -0,0 +1,13 @@
+﻿namespace LeetWars.Notifier.WebAPI.Models
+{
+    public class CodeRunResults
+    {
+        public long UserId { get; set; }
+        public long ChallengeVersionId { get; set; }
+        public string Language { get; set; } = string.Empty;
+        public bool IsBuilt { get; set; } = false;
```
Comment (pospyra) [inline backend/LeetWars.Notifier/LeetWars.Notifier.WebAPI/Models/CodeRunResults.cs]: the default value of a bool is false. i think u can remove it

File: backend/LeetWars.Builder/Services/XmlTestResultParserService.cs
Comment (AyTee001) [reply backend/LeetWars.Builder/Services/XmlTestResultParserService.cs:1]: I split methods, separated logic for creating a Test objects

File: frontend/src/app/modules/online-editor/online-editor-page/online-editor-page.component.ts (hunk 1 above)
Comment (AyTee001) [reply frontend/src/app/modules/online-editor/online-editor-page/online-editor-page.component.ts]: Removed

File: backend/LeetWars.Notifier/LeetWars.Notifier.WebAPI/Models/CodeRunResults.cs (hunk 2 above)
Comment (AyTee001) [reply backend/LeetWars.Notifier/LeetWars.Notifier.WebAPI/Models/CodeRunResults.cs]: Removed as well

File: backend/LeetWars.Builder/Services/XmlTestResultParserService.cs [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    {
+        public TestsOutput ParseCSharpTestResult(string xmlTestResult)
+        {
+            XmlDocument xmlDoc = new();
+
+            xmlDoc.LoadXml(xmlTestResult);
+
+            XmlNamespaceManager nsManager = new (xmlDoc.NameTable);
+
+            if(xmlDoc.DocumentElement?.NamespaceURI != null) nsManager.AddNamespace("ns", xmlDoc.DocumentElement.NamespaceURI);
```
Comment (pospyra) [inline backend/LeetWars.Builder/Services/XmlTestResultParserService.cs]: use braces

**Review by pospyra — APPROVED** [review]

File: backend/LeetWars.Builder/Interfaces/ISolutionRunnerService.cs [hunk 4]
```diff
@@ -1,7 +1,9 @@
-﻿namespace LeetWars.Builder.Interfaces
+﻿using Microsoft.AspNetCore.Mvc.Razor;
+
+namespace LeetWars.Builder.Interfaces
 {
     public interface ISolutionRunnerService
     {
-        Task RunCodeInContainerAsync(string imageName, string containerName);
+        Task<string> RunSolutionTestsAsync(string processName, string language, string code, string tests, string preloaded);
```
Comment (tatianahutii) [inline backend/LeetWars.Builder/Interfaces/ISolutionRunnerService.cs]: maybe one dto?

File: backend/LeetWars.Builder/LeetWars.Builder.csproj [hunk 5]
```diff
@@ -16,19 +16,17 @@
...
   </ItemGroup>
 
   <ItemGroup>
     <PackageReference Include="Docker.DotNet" Version="3.125.15" />
     <PackageReference Include="Microsoft.VisualStudio.Azure.Containers.Tools.Targets" Version="1.19.4" />
+    <PackageReference Include="SharpCompress" Version="0.33.0" />
   </ItemGroup>
 
   <ItemGroup>
+    <ProjectReference Include="..\LeetWars.Core\LeetWars.Core.Common\LeetWars.Core.Common.csproj" />
```
Comment (tatianahutii) [inline backend/LeetWars.Builder/LeetWars.Builder.csproj]: Need to add to Dockerfile

File: backend/LeetWars.Builder/Program.cs [hunk 6]
```diff
@@ -1,21 +1,27 @@
...
+builder.Services.AddTransient<ISolutionRunnerService, SolutionRunnerService>();
+
+builder.Services.AddTransient<ICodeRunManagerService, CodeRunManagerService>();
 
-//Test
-app.MapGet("/docker/csharp", () => new SolutionRunnerService().RunCodeInContainerAsync(DefaultRunnerImageNames.CSharpImage, "csharp-test-container"));
+builder.Services.AddSingleton<ITarManagementService, TarManagementService>();
 
-app.MapGet("/docker/js", () => new SolutionRunnerService().RunCodeInContainerAsync(DefaultRunnerImageNames.JSImage, "js-test-container"));
+builder.Services.AddSingleton<IXmlTestResultParserService, XmlTestResultParserService>();
```
Comment (tatianahutii) [inline backend/LeetWars.Builder/Program.cs]: separate extension method

File: backend/LeetWars.Builder/Services/ConsumeMessages.cs [hunk 7]
```diff
@@ -8,24 +13,43 @@ public class ConsumeMessages : BackgroundService
...
-                var message = Encoding.UTF8.GetString(body).Replace('"', '\0');
 
-                Console.WriteLine(message);
+                var message = Encoding.UTF8.GetString(body);
+
+                var request = JsonConvert.DeserializeObject<CodeRunRequest>(message);
+
+                var results = new CodeRunResults();
+
+                if (request != null && request.IsBuilt && (request.Tests != null || request.Tests != string.Empty))
```
Comment (tatianahutii) [inline backend/LeetWars.Builder/Services/ConsumeMessages.cs]: maybe use isNullOrEmpty?

File: backend/LeetWars.Builder/Services/SolutionRunnerService.cs [hunk 8]
```diff
@@ -1,29 +1,142 @@
...
         private readonly DockerClient _client = new DockerClientConfiguration().CreateClient();
 
-        public async Task RunCodeInContainerAsync(string imageName, string containerName)
+        private readonly IXmlTestResultParserService _parserService;
+
+        private readonly ITarManagementService _tarManagementService;
+
+        private readonly JsonSerializerSettings _serializerSettings = new() { ContractResolver = new CamelCasePropertyNamesContractResolver() };
+
+        private readonly string _localVolumeName = "LocalVolume";
```
Comment (tatianahutii) [inline backend/LeetWars.Builder/Services/SolutionRunnerService.cs:20]: appsettings

File: backend/LeetWars.Builder/Services/SolutionRunnerService.cs [hunk 9]
```diff
@@ -1,29 +1,142 @@
...
+            _tarManagementService = tarManagementService;
+        }
+
+        public async Task<string> RunSolutionTestsAsync(string processName, string language, string code, string tests, string preloaded)
+        {
+            return language switch
             {
-                Image = imageName,
-                Name = containerName
+                "csharp" => JsonConvert.SerializeObject(await RunCSharpSolutionTestsAsync(processName, code, tests, preloaded), _serializerSettings),
```
Comment (tatianahutii) [inline backend/LeetWars.Builder/Services/SolutionRunnerService.cs]: It would be better to create shared const and use them everywhere

File: backend/LeetWars.Builder/Services/SolutionRunnerService.cs [hunk 10]
```diff
@@ -1,29 +1,142 @@
...
+
+            await _client.Containers.WaitContainerAsync(container.ID);
+
+            var stringResult = await FileInContainerToStringAsync(container.ID, $"/{_localVolumeName}/{DefaultJSFileNaming.TestResultsFileName}");
+
+            await _client.Containers.RemoveContainerAsync(container.ID, new ContainerRemoveParameters { Force = true });
+
+            await _client.Volumes.RemoveAsync(volumeName);
+
+            return _parserService.ParseJSTestResult(stringResult);
```
Comment (tatianahutii) [inline backend/LeetWars.Builder/Services/SolutionRunnerService.cs]: the same as 57-67

File: backend/LeetWars.Builder/Services/SolutionRunnerService.cs [hunk 11]
```diff
@@ -1,29 +1,142 @@
...
+            return await _client.Containers.CreateContainerAsync(containerParams);
+        }
+
+        private async Task StringToFileInContainerAsync(string input, string newFileNameWithExtenstion, string containerId, string pathInContainer)
+        {
+            var bytes = _tarManagementService.SingleFileToTarBytes(input, newFileNameWithExtenstion);
+
+            using var tarStream = new MemoryStream(bytes);
+
+            var archivePath = pathInContainer;
```
Comment (tatianahutii) [inline backend/LeetWars.Builder/Services/SolutionRunnerService.cs]: do we need this extra var?

File: backend/LeetWars.Builder/Services/TarManagementService.cs [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+            {
+                using (TarWriter writer = new(ms, new TarWriterOptions(CompressionType.None, false)))
+                {
+                    using MemoryStream inputMs = new(Encoding.UTF8.GetBytes(input));
+
+                    writer.Write(newFileNameWithExtension, inputMs);
+                }
+
+                tarBytes = ms.ToArray();
+            }
```
Comment (tatianahutii) [inline backend/LeetWars.Builder/Services/TarManagementService.cs]: will be more readable use using without brackets

File: backend/LeetWars.Builder/Services/XmlTestResultParserService.cs [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+using System.Globalization;
+using System.Xml;
+
+namespace LeetWars.Builder.Services
+{
+    public class XmlTestResultParserService : IXmlTestResultParserService
+    {
+        public TestsOutput ParseCSharpTestResult(string xmlTestResult)
+        {
+            XmlDocument xmlDoc = new();
```
Comment (tatianahutii) [inline backend/LeetWars.Builder/Services/XmlTestResultParserService.cs]: why do not use "var" everywhere in this file for class definition?

File: backend/LeetWars.Builder/Services/XmlTestResultParserService.cs [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+            {
+                throw new ArgumentException("Incorrect xml test file supplied, cannot find needed data");
+            }
+
+            ICollection<Test> tests = new List<Test>();
+
+            foreach (XmlNode testNode in unitTestResultNodes)
+            {     
+                tests.Add(ParseSingleCSharpTest(testNode, nsManager));
+            }
```
Comment (tatianahutii) [inline backend/LeetWars.Builder/Services/XmlTestResultParserService.cs]: select?

File: backend/LeetWars.Builder/Services/XmlTestResultParserService.cs [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+
+            string? passed = countersNode.Attributes?["passed"]?.Value;
+
+            string? failed = countersNode.Attributes?["failed"]?.Value;
+
+            string? outcome = resultSummaryNode.Attributes?["outcome"]?.Value;
+
+            string? startTimeString = startTimeNode.Attributes?["start"]?.Value;
+
+            string? finishTimeString = startTimeNode.Attributes?["finish"]?.Value;
```
Comment (tatianahutii) [inline backend/LeetWars.Builder/Services/XmlTestResultParserService.cs]: maybe it would be more readable to  move this to a separate method and return just an class with all these fields

File: backend/LeetWars.Builder/Services/XmlTestResultParserService.cs [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+                nsManager.AddNamespace("ns", xmlDoc.DocumentElement.NamespaceURI);
+            }
+
+            XmlNode? countersNode = xmlDoc.SelectSingleNode("/ns:TestRun/ns:ResultSummary/ns:Counters", nsManager);
+
+            XmlNode? resultSummaryNode = xmlDoc.SelectSingleNode("/ns:TestRun/ns:ResultSummary", nsManager);
+
+            XmlNode? startTimeNode = xmlDoc.SelectSingleNode("/ns:TestRun/ns:Times", nsManager);
+
+            XmlNodeList? unitTestResultNodes = xmlDoc.SelectNodes("/ns:TestRun/ns:Results/ns:UnitTestResult", nsManager);
```
Comment (tatianahutii) [inline backend/LeetWars.Builder/Services/XmlTestResultParserService.cs]: the same here

File: backend/LeetWars.Builder/Services/XmlTestResultParserService.cs [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+
+            string? failed = countersNode.Attributes?["failed"]?.Value;
+
+            string? outcome = resultSummaryNode.Attributes?["outcome"]?.Value;
+
+            string? startTimeString = startTimeNode.Attributes?["start"]?.Value;
+
+            string? finishTimeString = startTimeNode.Attributes?["finish"]?.Value;
+
+            if (passed == null || failed == null || outcome == null || startTimeString == null || finishTimeString == null || unitTestResultNodes == null)
```
Comment (tatianahutii) [inline backend/LeetWars.Builder/Services/XmlTestResultParserService.cs]: and with DTO use something like .GetType().GetProperties().All(p => p != null); for checking all fields

File: backend/LeetWars.Builder/Services/XmlTestResultParserService.cs [hunk 18]
```diff
@@ -0,0 +1,164 @@
...
+            foreach (XmlNode testNode in unitTestResultNodes)
+            {     
+                tests.Add(ParseSingleCSharpTest(testNode, nsManager));
+            }
+
+            DateTime startTime = DateTime.Parse(startTimeString, CultureInfo.InvariantCulture);
+
+            DateTime finishTime = DateTime.Parse(finishTimeString, CultureInfo.InvariantCulture);
+
+            TimeSpan totalTime = finishTime - startTime;
```
Comment (tatianahutii) [inline backend/LeetWars.Builder/Services/XmlTestResultParserService.cs]: the same (separate method)

File: backend/LeetWars.Builder/Services/XmlTestResultParserService.cs [hunk 19]
```diff
@@ -0,0 +1,164 @@
...
+            string? duration = testSuits.Attributes?["time"]?.Value;
+
+            string? testsCount = testSuits.Attributes?["tests"]?.Value;
+
+            string? failed = testSuits.Attributes?["failures"]?.Value;
+
+            if(duration == null || testsCount == null || failed == null) 
+            {
+                throw new ArgumentException("Incorrect xml test file supplied, cannot find needed data");
+            }
```
Comment (tatianahutii) [inline backend/LeetWars.Builder/Services/XmlTestResultParserService.cs]: I believe you can create just some kind of generic method and reuse it in all similar places

File: backend/LeetWars.Builder/Services/XmlTestResultParserService.cs [hunk 20]
```diff
@@ -0,0 +1,164 @@
...
+            int passed = int.Parse(testsCount) - int.Parse(failed);
+
+            XmlNodeList? testNodes = xmlDoc.SelectNodes("//testcase") ?? throw new ArgumentException("Incorrect xml test file supplied, cannot find needed data");
+
+            ICollection<Test> tests = new List<Test>();
+
+            foreach (XmlNode testNode in testNodes) 
+            {
+                tests.Add(ParseSingleJSTest(testNode));
+            }
```
Comment (tatianahutii) [inline backend/LeetWars.Builder/Services/XmlTestResultParserService.cs]: see the previous coments

File: backend/LeetWars.Core/LeetWars.Core.WebAPI/appsettings.json [hunk 21]
```diff
@@ -1,6 +1,6 @@
 {
   "ConnectionStrings": {
-    "LeetWarsCoreDBConnection": "Server=localhost;Database=LeetWarsCoreDB;Trusted_Connection=True;"
+    "LeetWarsCoreDBConnection": "Server=(localdb)\\mssqllocaldb;Database=LeetWarsCoreDB;Trusted_Connection=True;"
```
Comment (tatianahutii) [inline backend/LeetWars.Core/LeetWars.Core.WebAPI/appsettings.json]: please return back =)

File: frontend/src/app/modules/online-editor/online-editor-page/online-editor-page.component.ts [hunk 22]
```diff
@@ -52,6 +59,16 @@ export class OnlineEditorPageComponent implements OnDestroy, OnInit {
     }
 
     ngOnInit() {
+        this.signalRService.start();
+
+        this.signalRService.listenMessages((message: string) => {
```
Comment (tatianahutii) [inline frontend/src/app/modules/online-editor/online-editor-page/online-editor-page.component.ts]: cannot you just use the interface here?

File: frontend/src/app/modules/online-editor/online-editor-page/online-editor-page.component.ts [hunk 23]
```diff
@@ -52,6 +59,16 @@ export class OnlineEditorPageComponent implements OnDestroy, OnInit {
...
+            const codeRunResults: CodeRunResults = JSON.parse(message);
+
+            if (codeRunResults.isBuilt && codeRunResults.testRunResults !== null && codeRunResults.testRunResults !== '') {
```
Comment (tatianahutii) [inline frontend/src/app/modules/online-editor/online-editor-page/online-editor-page.component.ts]:
```suggestion
            if (codeRunResults.isBuilt && codeRunResults.testRunResults) {
```

File: frontend/src/app/modules/online-editor/online-editor-page/online-editor-page.component.ts [hunk 24]
```diff
@@ -86,6 +103,16 @@ export class OnlineEditorPageComponent implements OnDestroy, OnInit {
         (<any>window).monaco.editor.setModelLanguage((<any>window).monaco.editor.getModels()[0], selectedLang);
     }
 
+    showTestResults(testOutput: string) {
+        const testResults: TestsOutput = JSON.parse(testOutput);
+
+        if (testResults.isSuccess) {
```
Comment (tatianahutii) [inline frontend/src/app/modules/online-editor/online-editor-page/online-editor-page.component.ts]: do we have some third option?

File: frontend/src/app/modules/online-editor/online-editor-page/online-editor-page.component.ts [hunk 25]
```diff
@@ -86,6 +103,16 @@ export class OnlineEditorPageComponent implements OnDestroy, OnInit {
         (<any>window).monaco.editor.setModelLanguage((<any>window).monaco.editor.getModels()[0], selectedLang);
     }
 
+    showTestResults(testOutput: string) {
```
Comment (tatianahutii) [inline frontend/src/app/modules/online-editor/online-editor-page/online-editor-page.component.ts]: maybe interface?

File: frontend/src/app/shared/models/code-run-results/code-run-results.ts [hunk 26]
```diff
@@ -0,0 +1,8 @@
+export interface CodeRunResults {
+    userId: number;
+    challengeVersionId: number;
+    isBuilt: boolean;
+    language: string;
+    buildErrorOutput: string | null;
+    testRunResults: string | null;
```
Comment (tatianahutii) [inline frontend/src/app/shared/models/code-run-results/code-run-results.ts]:
```suggestion
    buildErrorOutput?: string;
    testRunResults?: string;
```

File: frontend/src/app/shared/models/tests-output/test.ts [hunk 27]
```diff
@@ -0,0 +1,6 @@
+export interface Test {
+    testName: string;
+    isSuccess: boolean;
+    errorMessage: string | null;
```
Comment (tatianahutii) [inline frontend/src/app/shared/models/tests-output/test.ts]:
```suggestion
    errorMessage?: string ;
```

File: frontend/src/app/shared/models/tests-output/test.ts [hunk 28]
```diff
@@ -0,0 +1,6 @@
+export interface Test {
```
Comment (tatianahutii) [inline frontend/src/app/shared/models/tests-output/test.ts]:
```suggestion
export interface ITest {
```

File: frontend/src/app/shared/models/code-run-results/code-run-results.ts [hunk 29]
```diff
@@ -0,0 +1,8 @@
+export interface CodeRunResults {
```
Comment (tatianahutii) [inline frontend/src/app/shared/models/code-run-results/code-run-results.ts]:
```suggestion
export interface ICodeRunResults {
```

File: frontend/src/app/shared/models/code-run-request/code-run-request.ts [hunk 30]
```diff
@@ -0,0 +1,9 @@
+export interface CodeRunRequest {
+    userId: number;
+    challengeVersionId: number;
+    isBuilt: boolean;
+    language: string;
+    userCode: string;
+    preloaded: string | null;
+    tests: string | null;
+}
```
Comment (tatianahutii) [inline frontend/src/app/shared/models/code-run-request/code-run-request.ts]:
```suggestion
export interface ICodeRunRequest {
    userId: number;
    challengeVersionId: number;
    isBuilt: boolean;
    language: string;
    userCode: string;
    preloaded?: string;
    tests?: string ;
}
```

**Review by tatianahutii — CHANGES_REQUESTED** [review]

File: backend/LeetWars.Builder/Interfaces/ISolutionRunnerService.cs (hunk 4 above)
Comment (AyTee001) [reply backend/LeetWars.Builder/Interfaces/ISolutionRunnerService.cs]: Created

File: backend/LeetWars.Builder/LeetWars.Builder.csproj (hunk 5 above)
Comment (AyTee001) [reply backend/LeetWars.Builder/LeetWars.Builder.csproj]: I tried building and launching builder container and it worked and did not crash. Also, I noticed similar thing works for LeetWars.Core's dockerfile

File: backend/LeetWars.Builder/Program.cs (hunk 6 above)
Comment (AyTee001) [reply backend/LeetWars.Builder/Program.cs]: Separated

File: backend/LeetWars.Builder/Services/ConsumeMessages.cs (hunk 7 above)
Comment (AyTee001) [reply backend/LeetWars.Builder/Services/ConsumeMessages.cs]: Made a correction

File: backend/LeetWars.Builder/Services/SolutionRunnerService.cs (hunk 8 above)
Comment (AyTee001) [reply backend/LeetWars.Builder/Services/SolutionRunnerService.cs:20]: I changed the use of json serializer

File: backend/LeetWars.Builder/Services/SolutionRunnerService.cs (hunk 9 above)
Comment (AyTee001) [reply backend/LeetWars.Builder/Services/SolutionRunnerService.cs]: Separated to a file with consts

File: backend/LeetWars.Builder/Services/SolutionRunnerService.cs (hunk 10 above)
Comment (AyTee001) [reply backend/LeetWars.Builder/Services/SolutionRunnerService.cs]: Refactored as a separate method

File: backend/LeetWars.Builder/Services/SolutionRunnerService.cs (hunk 11 above)
Comment (AyTee001) [reply backend/LeetWars.Builder/Services/SolutionRunnerService.cs]: Removed

File: backend/LeetWars.Builder/Services/TarManagementService.cs (hunk 12 above)
Comment (AyTee001) [reply backend/LeetWars.Builder/Services/TarManagementService.cs]: Removed brackets

File: backend/LeetWars.Builder/Services/XmlTestResultParserService.cs (hunk 13 above)
Comment (AyTee001) [reply backend/LeetWars.Builder/Services/XmlTestResultParserService.cs]: Fixed declarations list XmlDocument xmlDoc = new();

File: backend/LeetWars.Builder/Services/XmlTestResultParserService.cs (hunk 15 above)
Comment (AyTee001) [reply backend/LeetWars.Builder/Services/XmlTestResultParserService.cs]: Separated to a separate method and created an according dto

File: backend/LeetWars.Builder/Services/XmlTestResultParserService.cs (hunk 14 above)
Comment (AyTee001) [reply backend/LeetWars.Builder/Services/XmlTestResultParserService.cs]: I noticed that XmlNodeList does not have Select() avaliable directly so to use Select it would be needed to copy the list to another type

File: backend/LeetWars.Builder/Services/XmlTestResultParserService.cs (hunk 16 above)
Comment (AyTee001) [reply backend/LeetWars.Builder/Services/XmlTestResultParserService.cs]: Fixed

File: backend/LeetWars.Builder/Services/XmlTestResultParserService.cs (hunk 17 above)
Comment (AyTee001) [reply backend/LeetWars.Builder/Services/XmlTestResultParserService.cs]: Created a method to check for nulls in properties that can be used for all entities

File: backend/LeetWars.Builder/Services/XmlTestResultParserService.cs (hunk 18 above)
Comment (AyTee001) [reply backend/LeetWars.Builder/Services/XmlTestResultParserService.cs]: Fixed

File: backend/LeetWars.Builder/Services/XmlTestResultParserService.cs (hunk 19 above)
Comment (AyTee001) [reply backend/LeetWars.Builder/Services/XmlTestResultParserService.cs]: Created

File: backend/LeetWars.Builder/Services/XmlTestResultParserService.cs (hunk 20 above)
Comment (AyTee001) [reply backend/LeetWars.Builder/Services/XmlTestResultParserService.cs]: I noticed that XmlNodeList does not have Select() avaliable directly so to use Select it would be needed to copy the list to another type

File: backend/LeetWars.Core/LeetWars.Core.WebAPI/appsettings.json (hunk 21 above)
Comment (AyTee001) [reply backend/LeetWars.Core/LeetWars.Core.WebAPI/appsettings.json]: Returned : )

File: frontend/src/app/modules/online-editor/online-editor-page/online-editor-page.component.ts (hunk 23 above)
Comment (AyTee001) [reply frontend/src/app/modules/online-editor/online-editor-page/online-editor-page.component.ts]: Fixed

File: frontend/src/app/modules/online-editor/online-editor-page/online-editor-page.component.ts (hunk 24 above)
Comment (AyTee001) [reply frontend/src/app/modules/online-editor/online-editor-page/online-editor-page.component.ts]: no) I fixed that

File: frontend/src/app/modules/online-editor/online-editor-page/online-editor-page.component.ts (hunk 25 above)
Comment (AyTee001) [reply frontend/src/app/modules/online-editor/online-editor-page/online-editor-page.component.ts]: Substituted with interface

File: frontend/src/app/modules/online-editor/online-editor-page/online-editor-page.component.ts (hunk 22 above)
Comment (AyTee001) [reply frontend/src/app/modules/online-editor/online-editor-page/online-editor-page.component.ts]: I thought SignalR sends messages as strings, so I will still need to parse messages

File: frontend/src/app/shared/models/code-run-results/code-run-results.ts (hunk 26 above)
Comment (AyTee001) [reply frontend/src/app/shared/models/code-run-results/code-run-results.ts]: Fixed

File: frontend/src/app/shared/models/tests-output/test.ts (hunk 27 above)
Comment (AyTee001) [reply frontend/src/app/shared/models/tests-output/test.ts]: Fixed

File: frontend/src/app/shared/models/tests-output/test.ts (hunk 28 above)
Comment (AyTee001) [reply frontend/src/app/shared/models/tests-output/test.ts]: Renamed

File: frontend/src/app/shared/models/code-run-results/code-run-results.ts (hunk 29 above)
Comment (AyTee001) [reply frontend/src/app/shared/models/code-run-results/code-run-results.ts]: Renamed

File: frontend/src/app/shared/models/code-run-request/code-run-request.ts (hunk 30 above)
Comment (AyTee001) [reply frontend/src/app/shared/models/code-run-request/code-run-request.ts]: Fixed

File: backend/LeetWars.Builder/LeetWars.Builder.csproj (hunk 5 above)
Comment (tatianahutii) [reply backend/LeetWars.Builder/LeetWars.Builder.csproj]: All references should be added to dockerfile, so they can work separatly

File: backend/LeetWars.Builder/Services/XmlTestResultParserService.cs [hunk 31]
```diff
@@ -0,0 +1,210 @@
...
+            string? executionTime = testNode.Attributes?["duration"]?.Value;
+
+            string? testName = testNode.Attributes?["testName"]?.Value;
+
+            string? testOutcome = testNode.Attributes?["outcome"]?.Value;
+
+            if (executionTime == null || testName == null || testOutcome == null)
+            {
+                throw new ArgumentException("Incorrect xml test file supplied, cannot find needed data");
+            }
```
Comment (tatianahutii) [inline backend/LeetWars.Builder/Services/XmlTestResultParserService.cs]: CheckForNoNulls ? =)

File: backend/LeetWars.Builder/appsettings.json [hunk 32]
```diff
@@ -6,11 +6,17 @@
     }
   },
   "AllowedHosts": "*",
-  "Rabbit": "amqp://guest:guest@localhost:5672",
+  "Rabbit": "amqp://guest:guest@rabbitmq:5672",
```
Comment (tatianahutii) [inline backend/LeetWars.Builder/appsettings.json]: please return it back

File: backend/LeetWars.Notifier/LeetWars.Notifier.WebAPI/Services/MessageConsumerService.cs [hunk 33]
```diff
@@ -7,23 +10,30 @@ namespace LeetWars.Notifier.WebAPI.Services
...
+
             var handler = new EventHandler<BasicDeliverEventArgs>((model, args) =>
             {
                 var body = args.Body.ToArray();
+
                 var message = Encoding.UTF8.GetString(body);
-                Console.WriteLine(message); //impement some logic here using signalR
-                _consumerService.SetAcknowledge(args.DeliveryTag, false);
+
+                _hubContext.Clients.All.BroadcastMessage(message);
```
Comment (tatianahutii) [inline backend/LeetWars.Notifier/LeetWars.Notifier.WebAPI/Services/MessageConsumerService.cs]: not to all, to one specific

File: backend/LeetWars.Notifier/LeetWars.Notifier.WebAPI/appsettings.json [hunk 34]
```diff
@@ -6,7 +6,7 @@
     }
   },
   "AllowedHosts": "*",
-  "Rabbit": "amqp://guest:guest@localhost:5672",
+  "Rabbit": "amqp://guest:guest@rabbitmq:5672",
```
Comment (tatianahutii) [inline backend/LeetWars.Notifier/LeetWars.Notifier.WebAPI/appsettings.json]: see the previous comments

File: frontend/src/app/modules/challenges/online-editor-page/online-editor-page.component.ts [hunk 35]
```diff
@@ -62,6 +79,35 @@ export class OnlineEditorPageComponent extends BaseComponent implements OnInit {
...
+            userCode: 'public class Solution\r\n{\r\n    public bool IsNumPrime(int num)\r\n    ' +
+                        '{\r\n        throw new Exception("Exception!!!");\r\n    }\r\n}\r\n',
+            preloaded: undefined,
+            tests: 'using NUnit.Framework;\r\n\r\n[TestFixture]\r\npublic class Tests\r\n{\r\n    ' +
+                    'private Solution? _solutionClass;\r\n\r\n    [SetUp]\r\n    ' +
+                    'public void Setup()\r\n    {\r\n        _solutionClass = new Solution();\r\n    }\r\n\r\n    ' +
+                    '[Test]\r\n    public void IsPrime_InputIs1_ReturnFalse()\r\n    ' +
+                    '{\r\n        var result = _solutionClass.IsNumPrime(2);\r\n\r\n        ' +
+                    'Assert.IsFalse(result, "1 should not be prime");\r\n    }\r\n}',
+        };
```
Comment (tatianahutii) [inline frontend/src/app/modules/challenges/online-editor-page/online-editor-page.component.ts]: to utils file

**Review by tatianahutii — CHANGES_REQUESTED** [review]

File: backend/LeetWars.Builder/Services/XmlTestResultParserService.cs (hunk 31 above)
Comment (AyTee001) [reply backend/LeetWars.Builder/Services/XmlTestResultParserService.cs]: Fixed

File: backend/LeetWars.Builder/appsettings.json (hunk 32 above)
Comment (AyTee001) [reply backend/LeetWars.Builder/appsettings.json]: Returned

File: backend/LeetWars.Notifier/LeetWars.Notifier.WebAPI/Services/MessageConsumerService.cs (hunk 33 above)
Comment (AyTee001) [reply backend/LeetWars.Notifier/LeetWars.Notifier.WebAPI/Services/MessageConsumerService.cs]: Implemented

File: backend/LeetWars.Notifier/LeetWars.Notifier.WebAPI/appsettings.json (hunk 34 above)
Comment (AyTee001) [reply backend/LeetWars.Notifier/LeetWars.Notifier.WebAPI/appsettings.json]: Returned as well

File: frontend/src/app/modules/challenges/online-editor-page/online-editor-page.component.ts (hunk 35 above)
Comment (AyTee001) [reply frontend/src/app/modules/challenges/online-editor-page/online-editor-page.component.ts]: It now depends on something that is not a constant

File: backend/LeetWars.Builder/LeetWars.Builder.csproj (hunk 5 above)
Comment (AyTee001) [reply backend/LeetWars.Builder/LeetWars.Builder.csproj]: Now there is no connection from Builder to Common

File: frontend/src/app/modules/challenges/online-editor-page/online-editor-page.component.ts [hunk 36]
```diff
@@ -62,6 +79,43 @@ export class OnlineEditorPageComponent extends BaseComponent implements OnInit {
...
+                userCode: 'public class Solution\r\n{\r\n    public bool IsNumPrime(int num)\r\n    ' +
+                        '{\r\n        throw new Exception("Exception!!!");\r\n    }\r\n}\r\n',
+                preloaded: undefined,
+                tests: 'using NUnit.Framework;\r\n\r\n[TestFixture]\r\npublic class Tests\r\n{\r\n    ' +
+                    'private Solution? _solutionClass;\r\n\r\n    [SetUp]\r\n    ' +
+                    'public void Setup()\r\n    {\r\n        _solutionClass = new Solution();\r\n    }\r\n\r\n    ' +
+                    '[Test]\r\n    public void IsPrime_InputIs1_ReturnFalse()\r\n    ' +
+                    '{\r\n        var result = _solutionClass.IsNumPrime(2);\r\n\r\n        ' +
+                    'Assert.IsFalse(result, "1 should not be prime");\r\n    }\r\n}',
+            };
```
Comment (tatianahutii) [inline frontend/src/app/modules/challenges/online-editor-page/online-editor-page.component.ts]: please get rid of hard-coded data and move his to utils file

File: frontend/src/app/shared/shared.module.ts [hunk 37]
```diff
@@ -75,7 +73,6 @@ import { ProgressBarComponent } from './components/progress-bar/progress-bar.com
         TabsModule,
         MarkdownModule,
         AngularMarkdownEditorModule,
-        InfiniteScrollModule,
```
Comment (tatianahutii) [inline frontend/src/app/shared/shared.module.ts]: why did you delete this?

File: frontend/src/app/app.module.ts [hunk 38]
```diff
@@ -25,7 +25,6 @@ import { AppRoutingModule } from './app-routing.module';
         ToastrModule.forRoot(),
         NgxChartsModule,
         SharedModule,
-        CoreModule,
```
Comment (tatianahutii) [inline frontend/src/app/app.module.ts:28]: the same

**Review by tatianahutii — CHANGES_REQUESTED** [review]

File: frontend/src/app/app.module.ts (hunk 38 above)
Comment (AyTee001) [reply frontend/src/app/app.module.ts:28]: There is CoreModule referenced, it is just that eariler it was mentioned twice and now once

File: frontend/src/app/shared/shared.module.ts (hunk 37 above)
Comment (AyTee001) [reply frontend/src/app/shared/shared.module.ts]: Returned

File: frontend/src/app/modules/challenges/online-editor-page/online-editor-page.component.ts (hunk 36 above)
Comment (AyTee001) [reply frontend/src/app/modules/challenges/online-editor-page/online-editor-page.component.ts]: Placed in utils
